File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
        // to simplify the logic, do no checking
        // just do a normal left-to-right pre-order 
        // and overwrite the array list index by index
        
        //keep track of the current level/index of the array
        // with a helper function
        
        // actually do a normal inorder to reduce the number of
        // times the array list needs to get resized. 
    public List<Integer> rightSideView(TreeNode root) {
        int maxDepth = findMaxDepth(root, 0);
        Integer[] arr = new Integer[maxDepth];
        dfs(root, 0, arr);
        return Arrays.asList(arr);
    }
    private int findMaxDepth(TreeNode root, int depth) {
        if (root == null) return depth;
        return Math.max(findMaxDepth(root.left, depth+1),
                        findMaxDepth(root.right, depth+1));
    }
    void dfs(TreeNode root, int level, Integer[] arr) {
        if (root == null) return;
        dfs(root.left, level + 1, arr);
        arr[level] = root.val;
        dfs(root.right, level + 1, arr);    
        return;
    }
}

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun rightSideView(root: TreeNode?): List<Int> {
        val height = height(root, 0)
        val arr = Array(height) {0}
        preorderFill(root, arr, 0)
        return arr.asList()
     }
    fun preorderFill(root: TreeNode?, arr: Array<Int>, level: Int) {
        if (root == null) return
        preorderFill(root.left, arr, level+1)
        arr[level] = root.`val`
        preorderFill(root.right, arr, level+1)
        return
    }
    fun height(root: TreeNode?, depth: Int): Int {
        if (root == null) return depth
        return Math.max(height(root.left, depth+1),
                        height(root.right, depth+1))
    }
}